cdap-app-fabric/src/main/java/co/cask/cdap/extension/AbstractExtensionLoader.java

```diff
@@ -0,0 +1,226 @@
+      public EXTENSION load(EXTENSION_TYPE key) throws Exception {
+        Map<EXTENSION_TYPE, EXTENSION> extensions = findExtensions(key);
+        // If there is none found in the ext dir, try to look it up from the CDAP system class ClassLoader.
+        // This is for the unit-test case, where extensions are part of the test dependency, hence in the
+        // unit-test ClassLoader.
+        // If no provider was found, returns the NOT_SUPPORTED_PROVIDER so that we won't search again for
+        // this program type.
+        // Cannot use null because LoadingCache doesn't allow null value
+        return Objects.firstNonNull(extensions.get(key), defaultExtension);
+      }
+    });
+  }
+
+  /**
+   * Finds the first extension from the given {@link ServiceLoader} that supports the specified key. If the key
+   * specified is {@code null}, returns all extensions from the extensions directory.
+   */
+  private Map<EXTENSION_TYPE, EXTENSION> findExtensions(@Nullable EXTENSION_TYPE key) {
+    Map<EXTENSION_TYPE, EXTENSION> extensions = new HashMap<>();
+    // A LoadingCache from extension directory to ServiceLoader
+    final LoadingCache<File, ServiceLoader<EXTENSION>> serviceLoaderCache = createServiceLoaderCache();
+    for (String dir : extDirs) {
+      File extDir = new File(dir);
+      if (!extDir.isDirectory()) {
+        continue;
+      }
+
+      // Each module would be under a directory of the extension directory
+      for (File moduleDir : DirUtils.listFiles(extDir)) {
+        if (!moduleDir.isDirectory()) {
+          continue;
+        }
+        // Try to find a provider that can support the given program type.
+        try {
+          extensions.putAll(findExtensions(serviceLoaderCache.getUnchecked(moduleDir), key));
+        } catch (Exception e) {
+          LOG.warn("Exception raised when loading an extension from {}. Extension ignored.", moduleDir, e);
+        }
+      }
+    }
+    // For unit tests, try to load the extensions from the system classloader. Only add those extensions which
+    // were not already loaded using the extension classloader.
+    for (Map.Entry<EXTENSION_TYPE, EXTENSION> entry : findExtensions(systemExtensionLoader, key).entrySet()) {
+      if (!extensions.containsKey(entry.getKey())) {
+        extensions.put(entry.getKey(), entry.getValue());
+      }
+    }
+    return extensions;
+  }
+
+  /**
+   * Returns the first available extension that support the specified key using the specified {@link ServiceLoader}.
+   * If the key is {@code null}, returns all the extensions in the extensions directory
+   */
+  private Map<EXTENSION_TYPE, EXTENSION> findExtensions(ServiceLoader<EXTENSION> serviceLoader,
+                                                        @Nullable EXTENSION_TYPE key) {
+    Map<EXTENSION_TYPE, EXTENSION> extensions = new HashMap<>();
+    for (EXTENSION provider : serviceLoader) {
+      Set<EXTENSION_TYPE> cacheKeys = getSupportedTypesForProvider(provider);
```

Comment:
I had kept it this way, so that it can even support the case when the caller does not know the extension type, which is the case when *all* extensions need to be loaded.